modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C

```diff
@@ -28,92 +30,111 @@ PowerLawCreepStressUpdate::validParams()
+  const GenericReal<is_ad> stress_delta = effective_trial_stress - _three_shear_modulus * scalar;
+  const GenericReal<is_ad> creep_rate_derivative =
+      -_coefficient * _three_shear_modulus * _n_exponent *
+      std::pow(stress_delta, _n_exponent - 1.0) * _exponential * _exp_time;
   return creep_rate_derivative * _dt - 1.0;
 }
 
+template <bool is_ad>
 Real
-PowerLawCreepStressUpdate::computeStrainEnergyRateDensity(
-    const MaterialProperty<RankTwoTensor> & stress,
-    const MaterialProperty<RankTwoTensor> & strain_rate)
+PowerLawCreepStressUpdateTempl<is_ad>::computeStrainEnergyRateDensity(
+    const GenericMaterialProperty<RankTwoTensor, is_ad> & stress,
+    const GenericMaterialProperty<RankTwoTensor, is_ad> & strain_rate)
 {
   if (_n_exponent <= 1)
     return 0.0;
 
   Real creep_factor = _n_exponent / (_n_exponent + 1);
 
-  return creep_factor * stress[_qp].doubleContraction((strain_rate)[_qp]);
+  return MetaPhysicL::raw_value(creep_factor * stress[_qp].doubleContraction((strain_rate)[_qp]));
 }
 
+template <bool is_ad>
 void
-PowerLawCreepStressUpdate::computeStressFinalize(const RankTwoTensor & plastic_strain_increment)
+PowerLawCreepStressUpdateTempl<is_ad>::computeStressFinalize(
+    const GenericRankTwoTensor<is_ad> & plastic_strain_increment)
 {
   _creep_strain[_qp] += plastic_strain_increment;
 }
 
+template <bool is_ad>
 void
-PowerLawCreepStressUpdate::resetIncrementalMaterialProperties()
+PowerLawCreepStressUpdateTempl<is_ad>::resetIncrementalMaterialProperties()
 {
   _creep_strain[_qp] = _creep_strain_old[_qp];
 }
 
+template <bool is_ad>
 bool
-PowerLawCreepStressUpdate::substeppingCapabilityEnabled()
+PowerLawCreepStressUpdateTempl<is_ad>::substeppingCapabilityEnabled()
 {
-  return getParam<bool>("use_substep");
+  return this->template getParam<bool>("use_substep");
 }
-
-template Real PowerLawCreepStressUpdate::computeResidualInternal<Real>(const Real &, const Real &);
+template class PowerLawCreepStressUpdateTempl<false>;
+template class PowerLawCreepStressUpdateTempl<true>;
+template Real PowerLawCreepStressUpdateTempl<false>::computeResidualInternal<Real>(const Real &,
+                                                                                   const Real &);
+template ADReal
+PowerLawCreepStressUpdateTempl<true>::computeResidualInternal<ADReal>(const ADReal &,
+                                                                      const ADReal &);
 template ChainedReal
-PowerLawCreepStressUpdate::computeResidualInternal<ChainedReal>(const Real &, const ChainedReal &);
+PowerLawCreepStressUpdateTempl<false>::computeResidualInternal<ChainedReal>(const Real &,
+                                                                            const ChainedReal &);
+template ChainedADReal
+PowerLawCreepStressUpdateTempl<true>::computeResidualInternal<ChainedADReal>(const ADReal &,
+                                                                             const ChainedADReal &);
```

Comment:
The main reason of this PR (beyond the obvious code cleanup) is to prove out this concept with the AD/nonAD and ChainedReal/Real templates. Does this section make sense to you @dschwen ?

Reply:
couldn't you simply replace that second layer of templating with `GenericChainedReal<is_ad>`?

Reply:
Oh, never mind. I didn't see those other combos

Reply:
Yeah...kinda guess in check here, so please see if this makes sense to you